### ISSUE TITLE
Internal exception in ruby-pg CI with truffleruby-head

### ISSUE THREAD

**Issue by larskanis** [oracle/truffleruby#3478]
It happens at every run [here in the pg specs](https://github.com/ged/ruby-pg/blob/c59ca8f8c0812a3f76e9f473ec85f49cf3c682e9/spec/pg/basic_type_map_for_results_spec.rb#L142) when returning a float NaN value.

This error is raised since several weeks in truffleruby-head "24.1.0-dev-3a920de7, like ruby 3.2.2, GraalVM CE Native [x86_64-linux]". It doesn't happen in truffleruby "23.1.2, like ruby 3.2.2, Oracle GraalVM Native [x86_64-linux]".

Here is a failing CI run: https://github.com/ged/ruby-pg/actions/runs/8115487309/job/22183459895#step:12:458

The output:
```ruby
truffleruby: an internal exception escaped out of the interpreter,
please report it to https://github.com/oracle/truffleruby/issues

dead handle 0xbad0000000198f0 (com.oracle.truffle.api.CompilerDirectives.ShouldNotReachHere)
	from com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere(CompilerDirectives.java:574)
	from com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere(CompilerDirectives.java:520)
	from org.truffleruby.cext.UnwrapNode$UnwrapNativeNode.raiseError(UnwrapNode.java:107)
	from org.truffleruby.cext.UnwrapNode$UnwrapNativeNode.unwrapTaggedObject(UnwrapNode.java:92)
	from org.truffleruby.cext.UnwrapNodeGen$UnwrapNativeNodeGen$Inlined.execute(UnwrapNodeGen.java:382)
	from org.truffleruby.cext.UnwrapNode.unwrapGeneric(UnwrapNode.java:286)
	from org.truffleruby.cext.UnwrapNodeGen$Inlined.execute(UnwrapNodeGen.java:156)
	from org.truffleruby.cext.CExtNodes$CallWithCExtLockAndFrameAndUnwrapNode.callWithCExtLockAndFrame(CExtNodes.java:261)
	from org.truffleruby.cext.CExtNodesFactory$CallWithCExtLockAndFrameAndUnwrapNodeFactory$CallWithCExtLockAndFrameAndUnwrapNodeGen.execute(CExtNodesFactory.java:564)
	from org.truffleruby.language.locals.WriteLocalVariableNode.execute(WriteLocalVariableNode.java:28)
	from org.truffleruby.language.RubyContextSourceNode.executeVoid(RubyContextSourceNode.java:23)
	from org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:35)
	from org.truffleruby.core.module.ModuleNodes$DefineMethodNode$CallMethodWithLambdaBody.execute(ModuleNodes.java:1373)
	from org.truffleruby.language.RubyLambdaRootNode.execute(RubyLambdaRootNode.java:84)
/home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext_ruby.rb:24:in `getvalue'
	from /home/runner/work/ruby-pg/ruby-pg/spec/pg/basic_type_map_for_results_spec.rb:142:in `block (5 levels) in <top (required)>'
	from /home/runner/work/ruby-pg/ruby-pg/spec/pg/basic_type_map_for_results_spec.rb:129:in `each'
	from /home/runner/work/ruby-pg/ruby-pg/spec/pg/basic_type_map_for_results_spec.rb:129:in `block (4 levels) in <top (required)>'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `instance_exec'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `block in run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `block in with_around_and_singleton_context_hooks'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `block in with_around_example_hooks'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `block in run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'
	from /home/runner/work/ruby-pg/ruby-pg/spec/helpers.rb:56:in `block in included'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:390:in `execute_with'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:628:in `block (2 levels) in run_around_example_hooks_for'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:627:in `run_around_example_hooks_for'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `with_around_example_hooks'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `with_around_and_singleton_context_hooks'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:259:in `run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:646:in `block in run_examples'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `map'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `run_examples'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:607:in `run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `block in run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `map'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `block in run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `map'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `map'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2091:in `with_suite_hooks'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:116:in `block in run_specs'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/reporter.rb:74:in `report'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:115:in `run_specs'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:89:in `run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:71:in `run'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:45:in `invoke'
	from /home/runner/.rubies/truffleruby-head/lib/gems/gems/rspec-core-3.13.0/exe/rspec:4:in `<top (required)>'
	from <internal:core> core/kernel.rb:378:in `load'
	from /home/runner/.rubies/truffleruby-head/bin/rspec:25:in `<main>'
```

**Comment by flavorjones** [comment]
I've seen similar failures in Nokogiri's test suite, though it appears to be sporadic for me.

Example test output: https://github.com/sparklemotion/nokogiri/actions/runs/8194425716/job/22410492610#step:7:114

Version: `truffleruby 23.1.2, like ruby 3.2.2, Oracle GraalVM Native [x86_64-linux]`


```
dead handle 0xbad000000023028 (com.oracle.truffle.api.CompilerDirectives.ShouldNotReachHere)
	from com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere(CompilerDirectives.java:574)
	from com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere(CompilerDirectives.java:520)
	from org.truffleruby.cext.UnwrapNode$UnwrapNativeNode.raiseError(UnwrapNode.java:107)
	from org.truffleruby.cext.UnwrapNode$UnwrapNativeNode.unwrapTaggedObject(UnwrapNode.java:92)
	from org.truffleruby.cext.UnwrapNodeGen$UnwrapNativeNodeGen$Inlined.executeAndSpecialize(UnwrapNodeGen.java:421)
	from org.truffleruby.cext.UnwrapNodeGen$UnwrapNativeNodeGen$Inlined.execute(UnwrapNodeGen.java:387)
	from org.truffleruby.cext.UnwrapNode.longToWrapper(UnwrapNode.java:270)
	from org.truffleruby.cext.UnwrapNodeGen$Inlined.executeAndSpecialize(UnwrapNodeGen.java:183)
	from org.truffleruby.cext.UnwrapNodeGen$Inlined.execute(UnwrapNodeGen.java:1[58](https://github.com/sparklemotion/nokogiri/actions/runs/8194425716/job/22410492610#step:7:59))
	from org.truffleruby.cext.CExtNodes$CallWithCExtLockAndFrameAndUnwrapNode.callWithCExtLockAndFrame(CExtNodes.java:258)
	from org.truffleruby.cext.CExtNodesFactory$CallWithCExtLockAndFrameAndUnwrapNodeFactory$CallWithCExtLockAndFrameAndUnwrapNodeGen.executeAndSpecialize(CExtNodesFactory.java:577)
	from org.truffleruby.cext.CExtNodesFactory$CallWithCExtLockAndFrameAndUnwrapNodeFactory$CallWithCExtLockAndFrameAndUnwrapNodeGen.execute(CExtNodesFactory.java:556)
	from org.truffleruby.language.locals.WriteLocalVariableNode.execute(WriteLocalVariableNode.java:28)
	from org.truffleruby.language.RubyNode.doExecuteVoid(RubyNode.java:64)
	from org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
	from org.truffleruby.core.module.ModuleNodes$DefineMethodNode$CallMethodWithLambdaBody.execute(ModuleNodes.java:1373)
	from org.truffleruby.language.RubyLambdaRootNode.execute(RubyLambdaRootNode.java:84)
/home/runner/.rubies/truffleruby-23.1.2/lib/truffle/truffle/cext_ruby.rb:23:in `parent'
	from /home/runner/work/nokogiri/nokogiri/test/xml/test_node_set.rb:[60](https://github.com/sparklemotion/nokogiri/actions/runs/8194425716/job/22410492610#step:7:61)2:in `block (4 levels) in <class:TestNodeSet>'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:237:in `block in each'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:236:in `upto'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:236:in `each'
```

**Comment by eregon** [comment]
`dead handle 0xbad... (com.oracle.truffle.api.CompilerDirectives.ShouldNotReachHere)` usually means that some `VALUE` field was not marked properly. TruffleRuby may call the marking functions more often or at different places than CRuby, CRuby calls them during GC and that's not possible on JVM so TruffleRuby calls them e.g. after returning from a method defined in C using `DATA_PTR`.
So it could be some field missed to be marked in the marking function, or it could be a missing `RB_GC_GUARD`.
The Ruby part of the backtrace usually gives some hint about which field/local variable it is about.

**Comment by flavorjones** [comment]
Interesting. The two most recent Nokogiri errors I have found ruby backtraces to tests that deal with duplicating nodes:

```
/home/runner/.rubies/truffleruby-23.1.2/lib/truffle/truffle/cext_ruby.rb:23:in `parent'
	from /home/runner/work/nokogiri/nokogiri/test/xml/test_node_set.rb:602:in `block (4 levels) in <class:TestNodeSet>'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:237:in `block in each'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:236:in `upto'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:236:in `each'
	from /home/runner/work/nokogiri/nokogiri/test/xml/test_node_set.rb:601:in `test_0002_wraps each node within a dup of the Node argument'
```

and

```
/home/runner/.rubies/truffleruby-head/lib/truffle/truffle/cext_ruby.rb:24:in `[]'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:237:in `block in each'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:236:in `upto'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/node_set.rb:236:in `each'
	from <internal:core> core/enumerable.rb:594:in `any?'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/document_fragment.rb:103:in `css'
	from /home/runner/work/nokogiri/nokogiri/lib/nokogiri/xml/searchable.rb:144:in `at_css'
	from /home/runner/work/nokogiri/nokogiri/test/xml/test_document_fragment.rb:304:in `test_dup_creates_mutable_tree'
```

I'll take a deeper look when I get a chance.

**Comment by ntkme** [comment]
@eregon It happens rarely. For most of the time it works just fine. I created a new issue here: https://github.com/oracle/truffleruby/issues/3500

**Comment by eregon** [comment]
I can reproduce the issue reliably in a C API spec, with both bignums and floats (`v = LONG2NUM(INT64_MAX);`, `v = DBL2NUM(0.0/0.0);`, `v = rb_eval_string("0.0/0.0")`).

**Comment by eregon** [comment]
The issue for the Float case seems that the ValueWrapper is not kept alive, after the `Init_` function has returned.
And when that GC's we lose the mapping from native address to the Float instance (a `java.lang.Double`).
`rb_global_variable()` etc add the relevant objects (e.g. the Float instance) in `GC_REGISTERED_ADDRESSES`. For regular Ruby objects they hold onto their ValueValue wrapper too, so that works fine, but for primitives like Float there is no way to store the ValueWrapper in a `java.lang.Double` instance.
So I think `GC_REGISTERED_ADDRESSES` should hold onto ValueWrapper's instead of the actual objects they refer to and that should fix it.

For bignums like `INT64_MAX` this actually fits in a Java `long` so is the same case as Float (cannot store a ValueWrapper in a `java.lang.Long` instance).
For "true" bignums that don't fit in a Java `long` it works fine already and it's like e.g. Symbols.
Fixnums are not affected because those are tagged pointers like in CRuby so VALUE->long is done only from the VALUE address and not needing any ValueWrapper/HandleBlock/etc.

A quick workaround until this is fixed is to run with `TRUFFLERUBYOPT="--experimental-options --keep-handles-alive"`.
That will keep all handles (`VALUE`) alive by leaking them, so obviously just a workaround but could be useful in CI until this is fixed.
We should have a fix very soon hopefully. It's surprising this didn't get caught by quite many CIs using truffleruby-dev, probably it's not that common to use Floats in an `Init_` function.

**Comment by flavorjones** [comment]
I've opened a new issue for the nokogiri errors above at https://github.com/oracle/truffleruby/issues/3503

**Comment by eregon** [comment]
This fix should be included for the 24.0.1 Release (Apr 16, 2024).
(and of course it's fixed on master and in `truffleruby-dev/head`)
